### PR TITLE
Python JSON parser: Ignore invalid enum string values if ignore_unknown_fields is set

### PR DESCRIPTION
**Pull request by antongrbin**
# Motivation

This PR fixes failing conformance tests for python with name `IgnoreUnknownEnumStringValue`.

The JSON parsing spec was discussed in https://github.com/protocolbuffers/protobuf/issues/7392.

Recent equivalent changes for other languages:
* Swift: https://github.com/apple/swift-protobuf/pull/1345
* C#: https://github.com/protocolbuffers/protobuf/pull/15758

# Changes

- 1st commit is a noop  refactoring to make relevant _ConvertScalarFieldValue invocations localized
- 2nd commit introduces the child exception of `ParseError` named `EnumStringValueParseError` which is suppressed if `ignore_unknown_fields` is set
- 3rd commit updates the conformance test failure lists

### CODE REVIEW

File: python/google/protobuf/internal/json_format_test.py [hunk 1]
```diff
@@ -1020,6 +1020,80 @@ def testParseEnumValue(self):
...
+  def testParseUnknownEnumStringValue_Scalar_Proto2(self):
+    message = json_format_pb2.TestNumbers()
+    text = '{"a": "UNKNOWN_STRING_VALUE"}'
+    json_format.Parse(text, message, ignore_unknown_fields=True)
+
+    self.assertFalse(message.HasField('a'))
+
+  def testParseErrorForUnknownEnumValue_ScalarWithoutIgnore_Proto2(self):
+    message = json_format_pb2.TestNumbers()
+    self.assertRaisesRegexp(
```
Comment (antongrbin) [inline python/google/protobuf/internal/json_format_test.py:1032]: Failure from CI:
https://github.com/protocolbuffers/protobuf/actions/runs/7993753108/job/21830301947?pr=15887#step:4:1822

File: src/google/protobuf/util/json_format.proto [hunk 2]
```diff
@@ -114,3 +115,11 @@ enum EnumValue {
 message TestDefaultEnumValue {
   optional EnumValue enum_value = 1 [default = DEFAULT];
 }
+
+message TestEnumMap {
```
Comment (antongrbin) [inline src/google/protobuf/util/json_format.proto:119]: Failure from CI:
- https://github.com/protocolbuffers/protobuf/actions/runs/7993753108/job/21830279412?pr=15887#step:13:3493
- https://github.com/protocolbuffers/protobuf/actions/runs/7993753108/job/21830264580?pr=15887#step:3:1046

File: python/google/protobuf/internal/json_format_test.py (hunk 1 above)
Comment (antongrbin) [reply python/google/protobuf/internal/json_format_test.py:1032]: Fixed by using `assertRaisesRegex` (without `p` at the end) which is consistent with the rest of the file.

File: src/google/protobuf/util/json_format.proto (hunk 2 above)
Comment (antongrbin) [reply src/google/protobuf/util/json_format.proto]: Fixed by renaming this message.

**Comment by antongrbin** [comment]
Rebased with upstream due to a conflict with an added conformance test (https://github.com/protocolbuffers/protobuf/commit/0fa67a94aacf4de1f1e35caca3389ed3e35a2cdf) that we now fixed as well. I used force-push since I don't believe review started already.

**Comment by antongrbin** [comment]
To over-communicate, this is now ready to review!

**Review by acozzette — APPROVED** [review]

**Comment by acozzette** [comment]
@antongrbin Sorry for the delay but would you mind rebasing your PR (or merging main into it)? There was a test run added just a few days ago and for technical reasons our sync process can't proceed until it sees that the test is passing.

**Comment by antongrbin** [comment]
@acozzette sure! Merged latest main into this branch. I am here if anything else is needed.

**Review by acozzette — APPROVED** [review]

**Comment by antongrbin** [comment]
@acozzette let me know if there is anything I can do with the CI failure for `Windows CMake shared`:

* `LINK : fatal error LNK1189: library limit of 65535 objects exceeded`
* https://github.com/protocolbuffers/protobuf/actions/runs/8216437017/job/22526672782#step:13:2714

**Comment by acozzette** [comment]
Sorry for the hassle but would you mind merging main into your branch one more time? I'm pretty sure that test failure is unrelated to your change, and it might have been fixed on main already.

**Comment by antongrbin** [comment]
@acozzette, done -- merged latest main here. No hassle at all -- I understand this is not a simple repo.

**Review by acozzette — APPROVED** [review]

**Comment by antongrbin** [comment]
@acozzette friendly ping :) Not that I am blocked on this (we are running a fork with this fix already), but would like to stop thinking about it sooner than later :)

**Comment by acozzette** [comment]
@antongrbin Sorry again for all the delays on this. I was able to get in a fix for the failing CMake test, so if you wouldn't mind merging main again then hopefully I can finally merge this.

**Comment by antongrbin** [comment]
> @antongrbin Sorry again for all the delays on this. I was able to get in a fix for the failing CMake test, so if you wouldn't mind merging main again then hopefully I can finally merge this.

@acozzette no worries! Merged with upstream :) Thank you for fixing the problem!!

**Comment by sbchisholm** [comment]
Is there a way now to use `ignore_unknown_fields` to ignore the unknown fields but still validate that the enum strings are valid options?

**Comment by acozzette** [comment]
No, for the purposes of JSON parsing an unknown enum string is considered an unknown field.

**Pull request closed without merging**